### DISCUSSION TITLE
Excessive indentation in code blocks when converting docx to markdown

### DISCUSSION THREAD

**Discussion by gitspeaks** [jgm/pandoc#9796] [Q&A]
I have a Word document containing some HTML, which I wrapped with `html` code block markers.

Example:

\`\`\`html
```
<!DOCTYPE html>
<html lang="en-US">
  <head>
  </head>
  <body>
  </body>
</html>
```
\`\`\`

I created a style in Word called “Source Code” and applied this style to the entire wrapped HTML. I then converted the docx to markdown using the following command:

```sh
pandoc -f docx -t markdown -o output.md input.docx
```

The entire code block is shifted by 4 spaces in the output file. I do not want this shift.

Any assistance with resolving this issue would be greatly appreciated.

**Comment by jgm** [comment]
Try using a `--reference-doc` and modifying the "Source Code" style.

**Reply by gitspeaks** [reply]
I followed the instructions in the [documentation](https://pandoc.org/MANUAL.html) to create the file

> To produce a custom reference.docx, first get a copy of the default reference.docx: pandoc -o custom-reference.docx --print-default-data-file reference.docx. Then open custom-reference.docx in Word, modify the styles as you wish, and save the file. For best results, do not make changes to this file other than modifying the styles used by pandoc

```
pandoc -o custom-reference.docx --print-default-data-file reference.docx
```

But there's no "Source Code" style in the output file (please see attached file). 
[custom-reference.docx](https://github.com/jgm/pandoc/files/15491350/custom-reference.docx)

I'm using pandoc 3.2

What's the correct approach to incorporate the "Source Code" style into the custom reference document without indentation ?

**Reply by TomBener** [reply]
It seems you need to generate a docx with the "Source Code" text to find it.

![CleanShot 2024-05-30 at 08 49 38@2x](https://github.com/jgm/pandoc/assets/49151155/b390c55c-9e3f-459e-b033-534e2b66d9da)

**Reply by gitspeaks** [reply]
I'm sorry, but I'm not sure I understand your suggestion. Could you please provide detailed steps on how to prepare a reference document with "Source Code" style that has no indentation?

Thank you!

**Reply by TomBener** [reply]
Save the following code block in `output.md`:

````html

Test source code style in docx:

```html
<!DOCTYPE html>
<html lang="en-US">
  <head>
  </head>
  <body>
  </body>
</html>
```
````

Then run the command to render `input.docx`:

```sh
pandoc -o output.md input.docx
```

Open the generated `input.docx` in MS Word and select the html code block within it, you'll see the "Code Block" in the Styles Pane on the right side. Click the style and edit it as you want, and then you can use the modified `input.docx` as the `reference-doc`.

**Reply by gitspeaks** [reply]
When I open the generated `input.docx` and examine the 'Source Code' style applied to the HTML block, I don't see any indentation as part of the style. I suspect this is added during docx-to-markdown conversion. Please see the attached reference document, source document, and the generated markdown file.

Here is how I did the conversion:

`pandoc -f docx -t markdown --wrap=none  source.docx -o traget.md --reference-doc=input.docx`

[traget.md](https://github.com/jgm/pandoc/files/15495388/traget.md)
[source.docx](https://github.com/jgm/pandoc/files/15495390/source.docx)
[input.docx](https://github.com/jgm/pandoc/files/15495391/input.docx)

**Reply by jgm** [reply]
I'm not seeing any indentation if I do, for example,
````
pandoc -o my.docx
```html
<!DOCTYPE html>
<html/>
```
^D
````
I notice that in your traget.md all the source is indented. That may be the problem.

**Comment by gitspeaks** [comment]
@jgm 
> I notice that in your traget.md all the source is indented. That may be the problem.

That's the problem and the whole point of this discussion  :)

**Reply by jgm** [reply]
Oh right, you're converting from docx to markdown.  Apologies, I misunderstood some of your original comment -- I had thought that the backticks were part of pandoc's output. But here's what I'm seeing:

```
% pandoc ~/Downloads/input.docx -t markdown
Test source code style in docx:

    <!DOCTYPE html>
    <html lang="en-US">
      <head>
      </head>
      <body>
      </body>
    </html>
```

That output is completely fine!  It's a different style of code block -- actually the original Markdown style.  The indentation indicates that it's a code block. Don't worry, the content will not be indented if you convert this to another format.

**Reply by jgm** [reply]
To be more explicit, there are two kinds of code blocks in pandoc's Markdown: indented code blocks (original Markdown style) and fenced code blocks.  The markdown writer uses indented code blocks when no language is specified.  (And this information isn't available in a docx source.)

**Reply by jgm** [reply]
Semantically,
````
    code
````
is equivalent to
````
```
code
```
````

**Reply by gitspeaks** [reply]
That's not my scenario. I *want* the backticks in the target.md markdown file (along with the programming language hint) *as they appear in the source.docx file*, just without the additional indentation. I'm not arguing against the semantic equivalence you described; it's just not my case.

**Reply by jgm** [reply]
OK, I thought input.docx was your input, but it's actually source.docx.
(Note that `--reference-doc` is irrelevant when you're *reading* docx.)
Anyway, my response is the same. In your source.docx, you have a code block that contains some backtick characters. The output markdown has a code block that contains those same backtick characters.  You seem to be expecting that markdown code block syntax will work in docx, which is an odd expectation. docx is not markdown.

**Comment by gitspeaks** [comment]
> You seem to be expecting that markdown code block syntax will work in docx, which is an odd expectation. docx is not markdown.

I don't expect the markdown code block syntax to work in docx, just as I don't expect the "Source Code" style to  "work" in docx. 
My understanding is that applying the "Source Code" style to some arbitrary text informs Pandoc that this text should be treated as a "block of code" during conversion. What this entails depends on the output format. For instance, when Pandoc is configured to convert to markdown, I expect it to check if the "Source Code" styled text starts and ends with backticks to identify the programming language, rather than simply treating them as part of the source code syntax.

If I don't include the backticks in the "Source Code" styled text, then I would need to add them in the target file (including the programming language hint) post-conversion, which would become a nightmare.

If that's too much to ask, how can I instruct Pandoc to output a specific Word paragraph "as-is" without any pre-processing during conversion? The reason I used the "Source Code" style initially is that, by default, < and > are correctly escaped in the output file, but this also results in HTML source code being escaped.

**Reply by gitspeaks** [reply]
Alternatively, if you are not comfortable with processing format-specific directives within the docx, then the "code-block" contract can be enhanced to include the programming language in the style name. For example, text styled with a style named "HTML Source Code" or "HTML Code Block" would instruct Pandoc to wrap the text with code markers and include the programming language hint in the output file when converting docx to markdown, perhaps using the first component in the style name for the language string.

**Reply by gitspeaks** [reply]
In any case, as mentioned, I would very much appreciate a way to instruct Pandoc not to encode or preprocess specific paragraphs during conversion and write the text "as-is" to the output file, effectively providing an escape hatch for those cases where "inline assembly" is required. This can be as simple as skipping encoding of text styled with a named style such as "Pandoc No Encoding" or similar wording.

**Reply by jgm** [reply]
>  For instance, when Pandoc is configured to convert to markdown, I expect it to check if the "Source Code" styled text starts and ends with backticks to identify the programming language, rather than simply treating them as part of the source code syntax.

Don't expect this.  That's expecting pandoc to be treating docx content as if it is markdown.  Backticks in a docx Source Code section are just backticks.

> If I don't include the backticks in the "Source Code" styled text, then I would need to add them in the target file (including the programming language hint) post-conversion, which would become a nightmare.

You could work around this by using a Lua filter and a simple convention. E.g. have each code block in the docx source start with a line containing `[language]`, and have your filter parse this, strip it out, and add the appropriate language as a class attribute.

> how can I instruct Pandoc to output a specific Word paragraph "as-is" without any pre-processing during conversion

There's no way to do this.  (And I'm sure you don't mean this literally ; you don't want a big block of OpenXML in your markdown, but that's what is in the docx container.)

**Comment by gitspeaks** [comment]
>You could work around this by using a Lua filter and a simple convention. E.g. have each code block in the docx source start with a line containing [language], and have your filter parse this, strip it out, and add the appropriate language as a class attribute.

Does the indentation occur before or after the Lua script executes? I don’t want to prepare the script only to end up back at square one.

**Reply by jgm** [reply]
The writer will produce an indented code block if there is no class indicating the language.  If the filter adds such a class, then the writer will produce a fenced code block.
